gpu/preconditioner/block_jacobi_kernels.cu

```diff
@@ -166,6 +166,210 @@ GKO_ENABLE_IMPLEMENTATION_SELECTION(select_apply, apply);
+    IndexType nnzlast = (tidx > 0) ? (row[tidx] - row[tidx - 1]) : 0;
+    v[tidx] = (nnzthis == nnzlast) ? 0 : 1;
+}
+
+template <typename IndexType>
+__global__ void compare_nzpattern_adjecent_rows(size_type m,
+                                                const IndexType *row,
+                                                const IndexType *col, bool *v)
+{
+    size_type tidx = blockDim.x * blockIdx.x + threadIdx.x;
+    size_type i = threadIdx.x;
+    int32 lastcol = -1;
+    int32 thiscol = -1;
+    int32 different = 0;
+    bool loc = 0;
+
+    if (tidx >= m) {
+        return;
+    }
+    if (v[tidx] == 0) {
+        size_type bound = row[tidx + 1] - row[tidx];
+        for (size_type j = 0; j < (bound + 32) / 32; j++) {
+            size_type ii = i + 32 * j;
+            if (ii < bound) {
+                thiscol = col[row[tidx] + ii];
+                lastcol = col[row[tidx - 1] + ii];
+                if (thiscol != lastcol) {
+                    different = 1;
+                }
+            }
+        }
+        loc = __any(different == 1);
+        if (i == 0) {
+            v[tidx] = loc;
+        }
+    }
+}
+
+
+template <typename IndexType>
+__global__ void generate_blockptr_natural_blocks(size_type num_rows,
+                                                 int32 max_block_size,
+                                                 const bool *v,
+                                                 IndexType *block_ptrs,
+                                                 size_type *num_blocks_arr)
+{
+    block_ptrs[0] = 0;
+    if (num_rows == 0) {
+        return;
+    }
+    size_type num_blocks = 1;
+    int32 current_block_size = 1;
+    for (size_type i = 1; i < num_rows; ++i) {
+        if (current_block_size < max_block_size && v[i] == 0) {
+            ++current_block_size;
+        } else {
+            block_ptrs[num_blocks] =
+                block_ptrs[num_blocks - 1] + current_block_size;
+            ++num_blocks;
+            current_block_size = 1;
+        }
+    }
+    block_ptrs[num_blocks] = block_ptrs[num_blocks - 1] + current_block_size;
+    num_blocks_arr[0] = num_blocks;
+}
+
+
+template <typename ValueType, typename IndexType>
+size_type find_natural_blocks(std::shared_ptr<const GpuExecutor> exec,
+                              const matrix::Csr<ValueType, IndexType> *mtx,
+                              int32 max_block_size, IndexType *block_ptrs)
+{
+    auto rows = mtx->get_num_rows();
+    auto cols = mtx->get_num_cols();
+    auto vals = mtx->get_const_values();
+    auto row_ptrs = mtx->get_const_row_ptrs();
+    auto col_idx = mtx->get_const_col_idxs();
+
+    Array<bool> d_array(exec, rows + 1);
+    auto d_v = d_array.get_data();
+
+    Array<size_type> d_nums_array(exec, 1);
+    auto d_nums = d_nums_array.get_data();
+
+    Array<size_type> nums_array(exec->get_master(), 1);
+    auto nums = nums_array.get_data();
+
+    constexpr int64 blocksize1_1 = 256;
```

Comment:
Probably good to create a `struct` and get these variables as members of the structs.